Massive display overhaul

harshil21 opened the pull request:
Biggest PR in airbrakes history. Provides a sleek and eye-catching TUI. It is recommended to use a modern terminal like kitty for the best experience (performance + graphics wise). 

There are still many more features, changes, and tests I would like to add, but for a first iteration, we have made it.

To try it out, simply run `uv run mock`. 

Closes #54.


Some non-display related changes, which was required for reliability reasons with the display:

- Forces multiprocessing start method to be "spawn" for all processes. "forkserver" does not work with `textual-image` unfortunately.
- `BaseIMU` now has a new shared variable called `request_to_start`, which is different from `is_running`, which is set only when the process is actually running. This distinction helps achieve cleaner starts and stops. 
- A new `ExtendedDataProcessor` class.
- It is currently not possible to run a `sim`. 


----
_Original post:_

To do:

- [x] Add T- launch time
- [x] Read all of the textual docs to understand how best to integrate and make the display
- [x] Barebones: Get display working with mock imu
- [x] Start adding graphs
- [x] Add launch file selector with some options, with ~picture of the rocket~ (the picture of the rocket is too low res for the terminal) along with relevant metadata
- [x] Add downrange distance calculation in the (x,y) plane in meters, it should show live trajectory, and estimated trajectory.
- [x] Add 3D model showing orientation of the rocket (not possible with textual, probably not worth using another library? Maybe a 2D model would have to suffice...)
- [ ] Detect drogue and main chute deployment
- [x] Add debug panel for information (apparently only possible with a 2nd terminal)
- [x] Integrate simulation speed into MockIMU
- [x] Create main and ~sim~ launch screen
- [x] Benchmark mode in mock - No display, just the main loop and show flight stats after benchmark.
- [x] Sound alarm if IMU has issues
- [ ] Document things
- [x] Make everything pretty
- [ ] Tests with textual

Estimated lines of code to add: 1000+ python, probably at least a couple hundred for the CSS.

harshil21 committed:
WIP: Start work on graphical ui

harshil21 committed:
Merge branch 'main' into term-gui

harshil21 committed:
Add T- launch time

harshil21 committed:
Merge branch nits and fix conflicts

harshil21 committed:
prints

harshil21 committed:
Merge main and fix conflicts

harshil21 committed:
WIP: add title

harshil21 committed:
Merge main and fix conflicts

harshil21 committed:
Use ISO 8601 format for representing date

harshil21 committed:
Fully fix T- launch time, read metadata

harshil21 committed:
Chunked reading of mock file for faster initial read

harshil21 committed:
Tiny bugfix in mock imu

harshil21 committed:
Move graphics folder out of mock

harshil21 committed:
Move landed_coordinates under 'flight_data' in metadata

harshil21 committed:
add flight length to metadata

harshil21 committed:
Initial work on startup display

harshil21 committed:
WIP: Start work on main flight display

harshil21 committed:
WIP: Almost done with basic flight display

harshil21 committed:
Design main app layout and finish DebugPanel

harshil21 committed:
Feat: Add ability to adjust sim speed

harshil21 committed:
Rename terminal.py -> application.py

harshil21 committed:
Split things into different files

harshil21 committed:
Add simulation time

There are performance issues when sim_speed < 1 with UI, even though you would expect that it should be the opposite... Need to debug why

harshil21 committed:
Disable real sim flight for now

harshil21 committed:
Add uv.lock and update pyproject.toml

harshil21 committed:
Fix performance issues by updating display from thread

harshil21 committed:
Fix a bug in time display

harshil21 committed:
Add initial cpu usage implementation

I think this will go into tabbed graphs section

harshil21 committed:
WIP: Rework cpu usage

harshil21 committed:
Finish CPU bar logic

Colors for higher usage pending

harshil21 committed:
Finish CPU widget logic and main styling

harshil21 committed:
Fix legacy main.py code

Just using this for speed comparisons

harshil21 committed:
Merge main and fix conflicts

harshil21 committed:
Properly merge main and get the replay working

harshil21 committed:
Merge main and fix conflicts

harshil21 committed:
Get mock working again, refactor launch_selector css

harshil21 committed:
Add camera CPUBar to GUI

harshil21 committed:
Merge main and fix conflicts!

harshil21 committed:
Fix display for latest textual version

harshil21 committed:
Add new display fields

harshil21 committed:
Fix performance regression in IMU

harshil21 committed:
Don't normalize CPU usage

harshil21 committed:
Feat: Flight Progress Bar

Also some cleanup:
1. Revert back to set_interval for updating telemetry, there is no noticeable performance penalty
2. Make mock logger a little more resilient if it has already been shutdown once and we are shutting it down again.
3. Correct the flight time in metadata for genesis launches with new state changes.

harshil21 committed:
WIP: Begin work on graphs

harshil21 committed:
Merge branch 'main' into term-gui

harshil21 committed:
Got tabbed graphs working!

harshil21 committed:
Fix no plot shown when sim is paused

harshil21 committed:
Make time blink when sim is paused

harshil21 committed:
Change state label to just the state

harshil21 committed:
Feat: Downrange Map

Also:
1. New ExtendedDataProcessor class for doing calculations which are only relevant in the mock.
2. Once again revert back to updating the GUI from the main loop, so it gets updated at a constant rate.
3. Update wrong times in metadata, and add the landed coords of legacy launch 2

harshil21 committed:
Fix nasty bug with incorrect downrange distance

harshil21 committed:
Add context manager for InformationStore

harshil21 committed:
Merge main and fix conflicts

harshil21 committed:
Add textual-image and bump other deps

harshil21 committed:
Beautification: Borders, colors, layout changes

Also added more stuff to ExtendedDataProcessor, and fix MockIMU get_launch_time()

harshil21 committed:
Merge branch 'main' into term-gui

harshil21 committed:
Add place name of launch site to metadata

harshil21 committed:
Beautification: Add picture to launch screen, use catpuccin-mocha theme

Also start standardizing colors

harshil21 committed:
Ignore repomix files

harshil21 committed:
Fix pitch calculation if IMU is differently oriented

harshil21 committed:
WIP: Start work on rocket visualization

Also some UI improvements

harshil21 committed:
WIP: Almost done with rocket visualization. It looks amazing!

Credit to grok for really understanding the assignment and nailing most of it. Also adds the actual image of purple launch (after removing the people in it using AI)

harshil21 committed:
WIP: Changing scale in rocket vis

harshil21 committed:
Merge branch 'main' into term-gui

harshil21 committed:
WIP: Change target apogee config

harshil21 committed:
Merge main and fix conflicts

harshil21 committed:
Pelican flight 4 changes + new deps

JacksonElia commented:
Can we get this merged sometime soon, even if unfinished? The PR is just getting too big, and I want to add a few things to the UI sometime.

harshil21 committed:
Merge main and fix conflicts

harshil21 committed:
Run docformatter, use old textual-pyfiglet version

Might need to fork the repository unfortunately

harshil21 committed:
Feat: Set a dynamic target apogee for every flight

harshil21 committed:
Update visualization to show airbrakes for MAX_EXTENSION also

harshil21 committed:
Some cleanup and better error handling for target apogee

harshil21 committed:
Feat: Allow skipping launch selector screen

harshil21 committed:
Feat: Ability to go back to launch select screen when running a mock

harshil21 committed:
Feat: Benchmark mode

This also changes up the architecture a little bit, and helps the introduction of uv run real interface. Moreover calculating average_acceleration was removed. Not sure why I added that in the first place. There was also a lot of cleanup

harshil21 committed:
Officially define a replay screen for more consistency and modularity

harshil21 committed:
Add slots to all textual classes

harshil21 committed:
Rewrite architecture in preparation for the real flight UI

harshil21 committed:
Add initial uv run real interface

harshil21 committed:
Bump textual-pyfiglet, use color gradient for airbrakes logo

harshil21 committed:
Feat: Keyboard shortcuts, tiny visual improvements

harshil21 committed:
Bump to v2.0.0

harshil21 committed:
Disable garbage collector for benchmarks

harshil21 committed:
Disable focus on some widgets to help with keyboard navigation

harshil21 committed:
Add launch picture so demo could work with uvx

I didn't test if it does work since the repo is now private anyway

harshil21 committed:
Performance improvements speically when running a real flight

harshil21 committed:
use real imu

harshil21 committed:
Fix real UI bugs

harshil21 committed:
Fix bug where the display would not stop upon motorburn

harshil21 committed:
Bump dependencies

JacksonElia reviewed (commented):
I will look at this when I have time, +5400 lines is scary 😭

harshil21 committed:
Bump textual-pyfiglet again

harshil21 committed:
Bump uv version and ruff versions in CI

harshil21 committed:
Adjust sleep comment

harshil21 committed:
Merge main and fix conflicts. Mock is NOT working as of this commit

harshil21 committed:
Fix the most frustating bug of my lifetime

harshil21 committed:
Remove the -d and -v arguments from the global parser

harshil21 committed:
Satisfy the ty type checker

harshil21 committed:
Use the textual-speedups library for some performance gains

harshil21 committed:
Get integration tests working and cut their time in half

harshil21 committed:
Add integrated altitude to the display

harshil21 committed:
small display performance improvement, fix wrong css

JacksonElia reviewed (commented):
just starting the massive review, good work

harshil21 committed:
Review: No magic numbers

harshil21 committed:
Bump dependencies

harshil21 committed:
Use custom font, fix ProgressBar bug, suppress ty warning

harshil21 committed:
Bump pre-commit versions

harshil21 committed:
Fix textual-plot's margin left on new version

harshil21 committed:
Bump textual-pyfiglet again, fix wrong css, make cpu bar look better in light theme

harshil21 committed:
Fix yet another dumb metadata typo

harshil21 committed:
Add json schema and pre-commit validation hook

harshil21 committed:
Start adding tests

harshil21 committed:
Bump dependencies - we can remove smblock.tlf install workaround

harshil21 committed:
Remove invalid data from legacy launch 1 data

harshil21 committed:
Fix a bug when running uv run real

harshil21 committed:
Merge branch 'main' into term-gui

harshil21 committed:
Merge pull request #214 from NCSU-High-Powered-Rocketry-Club/json-schema

Add a JSON schema for `metadata.json`

harshil21 committed:
Merge main and fix conflicts

harshil21 committed:
Merge main and fix conflicts

harshil21 committed:
Adapt this branch to new changes from main - no cam, no apogee conv finding

harshil21 committed:
Add uv.lock

harshil21 committed:
Merge main and fix conflicts

harshil21 committed:
Add government work picture

harshil21 committed:
Merge main and fix conflicts

harshil21 committed:
Rename jpg

harshil21 committed:
Actually add the jpg

harshil21 committed:
Add uv.lock back

harshil21 committed:
Merge main and fix conflicts

harshil21 committed:
Rename interest launch jpg

harshil21 committed:
Fix crash when opening graph for pred apogee

harshil21 committed:
Merge main and fix conflicts

Also add govt work 1 picture, remove cpu usage widget, remove mp workarounds, don't run json schema check because it fails for some reason

harshil21 committed:
Add govt work 2 data

harshil21 committed:
Clarify why we didn't hit the target apogee

harshil21 committed:
Fix pitch calc test

harshil21 committed:
Merge branch 'main' into term-gui

harshil21 committed:
Merge branch 'add-govt2' into term-gui

harshil21 committed:
Add govt work 2 data
